### PR TITLE
new_audit: add long-tasks diagnostic audit for surfacing top long tasks

### PR DESCRIPTION
**Pull request by Beytoven**
Addresses #10705

### CODE REVIEW

**Comment by connorjclark** [comment]
@brendankenny @patrickhulce I assume `main-thread-tasks` is for programmatic usages via HTTP Archive. true? Do we want to keep it? If so, @Beytoven we'll need to make a new audit instead `main-thread-long-tasks` ...

**Comment by patrickhulce** [comment]
Ooh, yeah good catch @connorjclark. It'd be a shame to lose this data. Not only for HA but it's a great diagnostic tool as well. I'm not sure if the visualizations in calibre and such use this or the trace itself.

Would you be able to create a `long-tasks` specific one @Beytoven ?

File: lighthouse-core/audits/long-tasks.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+const Audit = require('./audit.js');
+const NetworkRequest = require('../lib/network-request.js');
+const NetworkRecords = require('../computed/network-records.js');
+const i18n = require('../lib/i18n/i18n.js');
+const MainThreadTasks = require('../computed/main-thread-tasks.js');
+
+const UIStrings = {
+  /** Title of a diagnostic LH audit that provides details on the longest running tasks that occur when the page loads. */
+  title: 'Avoids long-running tasks',
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:16]: Since we're just listing the tasks here and not scoring the audit, title should be more declarative than prescriptive: `Long tasks` should be fine.

description could be `Lists the longest tasks on the main thread`

Do we need any sort of web.dev url for this? maybe just link to web.dev/tbt ?

File: lighthouse-core/audits/main-thread-tasks.js [hunk 2]
```diff
@@ -6,22 +6,8 @@
 'use strict';
```
Comment (connorjclark) [inline lighthouse-core/audits/main-thread-tasks.js]: revert these changes?

File: lighthouse-core/audits/long-tasks.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+   */
+  static async audit(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const tasks = await MainThreadTasks.request(trace, context);
+    const devtoolsLog = artifacts.devtoolsLogs[LongTasks.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    const jsURLs = LongTasks.getJavaScriptURLs(networkRecords);
+    const longtasks = [...tasks].sort((a, b) => b.duration - a.duration)
+      .filter(t => t.duration >= 50 && t.unbounded === false && !t.parent)
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:97]:
```suggestion
      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
```

Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:97]: filter, then sort (less work)

File: lighthouse-core/audits/long-tasks.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      .slice(0, 20);
+
+    const results = longtasks.map(t => ({
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:100]: t -> task

File: lighthouse-core/audits/long-tasks.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    return urls;
+  }
+
+  /**
+   * @param {LH.Artifacts.TaskNode} task
+   * @param {Set<string>} jsURLs
+   * @return {string}
+   */
+  static getAttributableURLForTask(task, jsURLs) {
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:69]: this is `BootupTime.getAttributableURLForTask` right? use that instead?

File: lighthouse-core/audits/long-tasks.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      url: LongTasks.getAttributableURLForTask(t, jsURLs),
+      group: t.group.label,
+      start: t.startTime,
+      self: t.selfTime,
+      duration: t.duration,
+    }));
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headings = [
+      {key: 'url', itemType: 'url', granularity: 1, text: str_(i18n.UIStrings.columnURL)},
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:110]: remove granularity: 1 for url

do we want 1 for start/duration? ms default is 10, which feels better than 1 (these numbers aren't nearly that precise)

File: lighthouse-core/lib/i18n/locales/uk.json [hunk 7]
```diff
@@ -1136,12 +1136,6 @@
   "lighthouse-core/audits/timing-budget.js | title": {
     "message": "Обмеження часу"
   },
-  "lighthouse-core/audits/user-timings.js | columnDuration": {
```
Comment (connorjclark) [inline lighthouse-core/lib/i18n/locales/uk.json:1139]: FYI we will need a string import when this lands @exterkamp

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+const LongTasks = require('../../audits/long-tasks.js');
+
+const acceptableTrace = require('../fixtures/traces/progressive-app-m60.json');
+const devtoolsLog = require('../fixtures/traces/progressive-app-m60.devtools.log.json');
+
+/* eslint-env jest */
+
+describe('Long tasks audit', () => {
+  it('should filter out top level tasks with a duration over 50 ms', async () => {
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:16]:
```suggestion
  it('should filter out top level tasks shorter than 50 ms', async () => {
```

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      traces: {defaultPass: acceptableTrace},
+      devtoolsLogs: {defaultPass: devtoolsLog},
+    };
+    const result = await LongTasks.audit(artifacts, {computedCache: new Map()});
+    expect(result.details.items).toHaveLength(4);
+
+    for (const item of result.details.items) {
+      expect(Number.isFinite(item.start)).toBeTruthy();
+      expect(Number.isFinite(item.duration)).toBeTruthy();
+      expect(item.duration).toBeGreaterThan(50);
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:27]: greater than or equal

File: lighthouse-core/audits/long-tasks.js (hunk 1 above)
Comment (Beytoven) [reply lighthouse-core/audits/long-tasks.js:16]: I thought about linking to tbt, but unsure as I don't want this audit to be seen as specifically for tbt.

**Comment by connorjclark** [comment]
can you add a `displayValue` that says "1 long task found" / "3 long tasks found" etc., like for the layout shift audit?

**Comment by Beytoven** [comment]
> can you add a `displayValue` that says "1 long task found" / "3 long tasks found" etc., like for the layout shift audit?
> 
> after that I think we're good here

Done.

File: lighthouse-core/audits/long-tasks.js [hunk 10]
```diff
@@ -72,6 +77,7 @@ class LongTasks extends Audit {
     return {
       score: 1,
       details: tableDetails,
+      displayValue: str_(UIStrings.displayValue, {itemCount: results.length}),
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:80]: some audits dont set display value for zero items:

https://github.com/GoogleChrome/lighthouse/blob/f3d0e3459d8fd15b055148dec0ae4e430df6495b/lighthouse-core/audits/deprecations.js#L91

should we do the same here and the CLS audit?

Comment (Beytoven) [reply lighthouse-core/audits/long-tasks.js:80]: I've gone back and forth on it. I don't have a strong preference either way. For consistency, I can go ahead and do the same for these new audits.

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    const result = await LongTasks.audit(artifacts, {computedCache: new Map()});
+    expect(result.details.items).toHaveLength(4);
+    expect(result.displayValue).toBeDisplayString('4 long tasks found');
+
+    for (const item of result.details.items) {
+      expect(Number.isFinite(item.start)).toBeTruthy();
+      expect(Number.isFinite(item.duration)).toBeTruthy();
+      expect(item.duration).toBeGreaterThanOrEqual(50);
+    }
+  });
```
Comment (brendankenny) [inline lighthouse-core/test/audits/long-tasks-test.js:30]: did y'all discuss more tests? It would be good to hit a few more things in there, unbounded tasks, unattributable URLs, etc. Those situations may already be in our existing fixtures (but will take a little digging to make sure they're in there), and there's [`create-test-trace.js`](https://github.com/GoogleChrome/lighthouse/blob/aaa6f7455cc415151759c52e741becac97a2ccde/lighthouse-core/test/create-test-trace.js) which can help make a simple valid-ish trace with certain top-level task characteristics.

Also, smoke test(s)! They might be tricky to avoid flakiness, but there are enough intentionally constructed perf/metric tests in there that at least some things are stable enough to assert.

File: lighthouse-core/audits/long-tasks.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+const i18n = require('../lib/i18n/i18n.js');
+const MainThreadTasks = require('../computed/main-thread-tasks.js');
+const BootupTime = require('./bootup-time.js');
+
+const UIStrings = {
+  /** Title of a diagnostic LH audit that provides details on the longest running tasks that occur when the page loads. */
+  title: 'Long main thread tasks',
+  /** Description of a diagnostic LH audit that tells the user to minimize the amount of long-running tasks on a page. */
+  description: 'Lists the longest tasks on the main thread, ' +
+    'useful for identifying worst contributors to input delay.',
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/audits/long-tasks.js]: I agree with this. We don't need a specific LH-url for this audit, the one you linked to looks good

File: lighthouse-core/audits/long-tasks.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      start: task.startTime,
+      self: task.selfTime,
+      duration: task.duration,
+    }));
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headings = [
+      {key: 'url', itemType: 'url', text: str_(i18n.UIStrings.columnURL)},
+      {key: 'start', itemType: 'ms', granularity: 10, text: str_(i18n.UIStrings.columnStartTime)},
+      {key: 'duration', itemType: 'ms', granularity: 10, text: str_(i18n.UIStrings.columnDuration)},
```
[the remark this answers is not in the document]
Comment (Beytoven) [reply lighthouse-core/audits/long-tasks.js]: Should we just update that audit to use a granularity of 10 instead? Is there a reason ms needs to be so precise in this case when the default is 10?

File: lighthouse-core/audits/long-tasks.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+const UIStrings = {
+  /** Title of a diagnostic LH audit that provides details on the longest running tasks that occur when the page loads. */
+  title: 'Long main-thread tasks',
+  /** Description of a diagnostic LH audit that shows the user the longest running tasks that occur when the page loads. */
+  description: 'Lists the longest tasks on the main thread, ' +
+    'useful for identifying worst contributors to input delay. ' +
+    '[Learn more](https://web.dev/long-tasks-devtools)',
+  /** [ICU Syntax] Label identifying the number of long-running CPU tasks that occurred while loading a web page. */
+  displayValue: `{itemCount, plural,
+  =1 {1 long task found}
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:23]: @exterkamp what were you saying this should be instead based on #10786 ?

File: lighthouse-core/audits/long-tasks.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    const tasks = await MainThreadTasks.request(trace, context);
+    const devtoolsLog = artifacts.devtoolsLogs[LongTasks.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+    const multiplier = settings.throttlingMethod === 'simulate' ?
+      settings.throttling.cpuSlowdownMultiplier : 1;
+
+    const jsURLs = BootupTime.getJavaScriptURLs(networkRecords);
+    // Only consider up to 20 long, top-level (no parent) tasks that have an explicit endTime
+    const longtasks = [...tasks]
+      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:61]:
```suggestion
    const longtasks = [...tasks]
      // account for simulated throttling *before* we filter to long tasks
      .map(t => ({...t, duration: t.duration * multiplier})
      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
```

File: lighthouse-core/audits/long-tasks.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const tasks = await MainThreadTasks.request(trace, context);
+    const devtoolsLog = artifacts.devtoolsLogs[LongTasks.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    const jsURLs = BootupTime.getJavaScriptURLs(networkRecords);
+    const longtasks = [...tasks]
+      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
+      .sort((a, b) => b.duration - a.duration)
+      .slice(0, 20);
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/audits/long-tasks.js:63]: startTime is going to be really tricky. the best way is going to be grabbing the pessimistic TTI graph and finding the startTime of the matching event id. 

That's a pain though, do we need to show it at first? It's not like the simulated start time is going to be all that meaningful for tracking it down in the trace either since it's a completely different execution path :/

File: lighthouse-core/audits/long-tasks.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    const tableDetails = Audit.makeTableDetails(headings, results);
+
+    let displayValue;
+    if (results.length > 0) {
+      displayValue = str_(UIStrings.displayValue, {itemCount: results.length});
+    }
+
+    return {
+      score: 1,
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:88]: for informative audits we typically `return {notApplicable: true}` when there's nothing to show otherwise they'll show up when there's no action to take.

File: lighthouse-core/audits/long-tasks.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+const Audit = require('./audit.js');
+const NetworkRecords = require('../computed/network-records.js');
+const i18n = require('../lib/i18n/i18n.js');
+const MainThreadTasks = require('../computed/main-thread-tasks.js');
+const BootupTime = require('./bootup-time.js');
+
+const UIStrings = {
+  /** Title of a diagnostic LH audit that provides details on the longest running tasks that occur when the page loads. */
+  title: 'Long main-thread tasks',
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:16]: @connorjclark I have some bad news 😬 

I'd have to disagree with the conclusion in https://github.com/GoogleChrome/lighthouse/pull/10736#discussion_r423389377 based on informativeness. We do the exact same thing of "just showing" the critical request chains but everything that is listed is *bad* and shouldn't be there so we still use a prescriptive title of "Avoid chaining critical requests", even if it's not scored. I liked `Avoids long main-thread tasks` :)

File: lighthouse-core/test/audits/long-tasks-test.js (hunk 11 above)
Comment (patrickhulce) [reply lighthouse-core/test/audits/long-tasks-test.js:30]: +1 to more test cases around here, specifically around the differences between provided/devtools throttling and simulated as well as something that would catch the fact that simulated tasks need to be multiplied *before* they're filtered out.

for non-flaky smokes, you should be able to manually construct particularly long, long tasks and then assert the first N are *at least* as long as you expect them. poor CI will probably create a few more on their own, but they should hopefully be just above the threshold and not creep into the longest long tasks positions

File: lighthouse-core/audits/long-tasks.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/audits/long-tasks.js:63]: uhh why isn't it just multiplier * startTime?

[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/audits/long-tasks.js:63]: yup :)

not only that but some tasks might also depend on a series of network requests that do not scale according to the `multiplier`

[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/audits/long-tasks.js:63]: I'm ok with leaving out start time for now. We'll need it for a future pass, if we want to visualize as a timeline.

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 19]
```diff
@@ -145,62 +169,41 @@ module.exports = [
...
-    },
-  },
-  {
-    lhr: {
-      requestedUrl: 'http://localhost:10200/perf/fonts.html',
-      finalUrl: 'http://localhost:10200/perf/fonts.html',
-      audits: {
-        'font-display': {
-          score: 0,
+        // TODO: uncomment when Chrome m83 lands
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:178]: should be fine now

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    };
+    const result = await LongTasks.audit(artifacts, {computedCache: new Map()});
+    expect(result.score).toBe(0);
+    expect(result.details.items).toHaveLength(4);
+    expect(result.displayValue).toBeDisplayString('4 long tasks found');
+    expect(result.notApplicable).toBeFalsy();
+
+    for (const item of result.details.items) {
+      expect(Number.isFinite(item.duration)).toBeTruthy();
+      expect(item.duration).toBeGreaterThanOrEqual(50);
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/long-tasks-test.js:77]:
```suggestion
      expect(item.duration).toEqual(200);
```

at least one explicit test for not using multiplier :)

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    expect(result.displayValue).toBeDisplayString('4 long tasks found');
+
+    for (const item of result.details.items) {
+      expect(Number.isFinite(item.duration)).toBeTruthy();
+      expect(item.duration).toBeGreaterThanOrEqual(50);
+      expect(item.url).toEqual('Unattributable');
+    }
+  });
+
+  it('should populate url when tasks have an attributable url', async () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/long-tasks-test.js:133]: nice touch :)

File: lighthouse-core/audits/long-tasks.js [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+      .map(t => ({...t, duration: t.duration * multiplier}))
+      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
+      .sort((a, b) => b.duration - a.duration)
+      .slice(0, 20);
+
+    const results = longtasks.map(task => ({
+      url: BootupTime.getAttributableURLForTask(task, jsURLs),
+      group: task.group.label,
+      start: task.startTime,
+      self: task.selfTime,
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:70]: I'm not sure selfTime is worth including for the toplevel task since anything there is just overhead.

If we include it should we ensure it matches the multiplier too and throw a test on it?

File: lighthouse-core/audits/long-tasks.js [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    const longtasks = [...tasks]
+      .map(t => ({...t, duration: t.duration * multiplier}))
+      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
+      .sort((a, b) => b.duration - a.duration)
+      .slice(0, 20);
+
+    const results = longtasks.map(task => ({
+      url: BootupTime.getAttributableURLForTask(task, jsURLs),
+      group: task.group.label,
+      start: task.startTime,
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:69]: I think it's going to be weird that `startTime` is reported on the actual timeline while `duration` is going to be on the simulated timeline, should we just exclude it for now?

File: lighthouse-core/audits/long-tasks.js [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+    // Only consider up to 20 long, top-level (no parent) tasks that have an explicit endTime
+    const longtasks = [...tasks]
+      .map(t => ({...t, duration: t.duration * multiplier}))
+      .filter(t => t.duration >= 50 && !t.unbounded && !t.parent)
+      .sort((a, b) => b.duration - a.duration)
+      .slice(0, 20);
+
+    const results = longtasks.map(task => ({
+      url: BootupTime.getAttributableURLForTask(task, jsURLs),
+      group: task.group.label,
```
Comment (patrickhulce) [inline lighthouse-core/audits/long-tasks.js:68]: don't know if `group` is worth including when it's value is *always* `Other`, maybe we exclude it?

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 25]
```diff
@@ -145,62 +169,35 @@ module.exports = [
                 node: {
                   type: 'node',
                   nodeLabel: 'img',
-                  path: '2,HTML,1,BODY,0,IMG',
+                  path: '0,HTML,1,BODY,0,DIV,0,IMG',
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:172]: I'd prefer removing `path` and instead adding something like `src` or another unique attribute here (can we match for that in snippet?)

On a quest to reduce paths in smoke tests :) https://github.com/GoogleChrome/lighthouse/pull/10827/files

File: lighthouse-core/audits/long-tasks.js [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+
+    const tableDetails = Audit.makeTableDetails(headings, results);
+
+    let displayValue;
+    if (results.length > 0) {
+      displayValue = str_(UIStrings.displayValue, {itemCount: results.length});
+    }
+
+    return {
+      score: Number(results.length === 0),
```
Comment (connorjclark) [inline lighthouse-core/audits/long-tasks.js:86]:
```suggestion
      score: results.length === 0 ? 1 : 0,
```

easier to read

File: lighthouse-core/lib/i18n/i18n.js [hunk 27]
```diff
@@ -84,6 +84,10 @@ const UIStrings = {
   columnOverBudget: 'Over Budget',
   /** Label for a column in a data table; entries will be a representation of a DOM element. */
   columnElement: 'Element',
+  /** Label for a column in a data table; entries will be the number of milliseconds since the page started loading. */
+  columnStartTime: 'Start Time',
```
Comment (connorjclark) [inline lighthouse-core/lib/i18n/i18n.js:88]: technically not needed here but I say keep it b/c it will be needed here in your followup

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 28]
```diff
@@ -145,62 +169,33 @@ module.exports = [
...
+                url: 'http://localhost:10200/perf/delayed-element.js',
+                duration: '>500',
+              },
+            },
+          },
+        },
       },
     },
   },
-  // TODO: uncomment when Chrome m83 lands
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:173]: I'm ok with removing this but can you make an issue + assign to yourself so we don't forget to add it back when m84 (comment is wrong) lands?

File: lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js [hunk 29]
```diff
@@ -136,6 +136,30 @@ module.exports = [
...
+    },
+  },
+  {
+    lhr: {
+      requestedUrl: 'http://localhost:10200/perf/trace-elements.html',
+      finalUrl: 'http://localhost:10200/perf/trace-elements.html',
+      audits: {
         'largest-contentful-paint-element': {
           score: null,
           displayValue: '1 element found',
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/test-definitions/perf/expectations.js:165]: driveby: can you remove `displayValue`?

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+        {ts: 3000, duration: 25},
+        {ts: 4000, duration: 50},
+      ],
+    });
+    const artifacts = {
+      traces: {defaultPass: trace},
+      devtoolsLogs: {defaultPass: devtoolsLog},
+    };
+    const result = await LongTasks.audit(artifacts, {computedCache: new Map()});
+    // expect(result.score).toBe(0);
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:96]: remove

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+/* eslint-env jest */
+
+/**
+ * @param {Number} count
+ * @param {Number} duration
+ * @param {Boolean} withChildTasks
+ */
+function generateTraceWithLongTasks(
+  count,
+  duration = 200,
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:23]: styling nit: we tend to only have line breaks in params if it is too long.

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 32]
```diff
@@ -0,0 +1,150 @@
...
+  withChildTasks = false
+) {
+  const baseTs = 1000;
+  const traceTasks = [];
+  for (let i = 1; i <= count; i++) {
+    const ts = baseTs * i;
+    const task = {ts, duration};
+    task.children = [];
+    if (withChildTasks) {
+      task.children.push({
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:33]: Can you add a comment detailing (or maybe an asci chart) the structure of the child tasks?

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 33]
```diff
@@ -0,0 +1,150 @@
...
+ */
+function generateTraceWithLongTasks(
+  count,
+  duration = 200,
+  withChildTasks = false
+) {
+  const baseTs = 1000;
+  const traceTasks = [];
+  for (let i = 1; i <= count; i++) {
+    const ts = baseTs * i;
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:29]: looks like tasks start at 1s instead of 0, is that important?

File: lighthouse-core/test/audits/long-tasks-test.js [hunk 34]
```diff
@@ -0,0 +1,150 @@
...
+
+    for (const item of result.details.items) {
+      expect(Number.isFinite(item.duration)).toBeTruthy();
+      expect(item.duration).toEqual(90);
+      expect(item.url).toEqual('Unattributable');
+    }
+  });
+
+  it('should populate url when tasks have an attributable url', async () => {
+    const trace = generateTraceWithLongTasks(1, 300, true);
```
Comment (connorjclark) [inline lighthouse-core/test/audits/long-tasks-test.js:134]: maybe use object parameter? can't tell what these values mean w/o looking at function.

File: lighthouse-core/test/audits/long-tasks-test.js (hunk 33 above)
Comment (Beytoven) [reply lighthouse-core/test/audits/long-tasks-test.js:29]: For this case, yes. I want to avoid any overlapping trace events so that no unintentional parent/child task relationships get formed since we need top level tasks for this audit. So starting at 1s avoids any of the added tasks overlapping with the required tasks defined in `CreateTestTrace()`.

File: lighthouse-core/test/audits/long-tasks-test.js (hunk 32 above)
Comment (Beytoven) [reply lighthouse-core/test/audits/long-tasks-test.js:33]: Good idea, will do.
